### DISCUSSION TITLE
TypeVars won't bind to class variable defaults?

### DISCUSSION THREAD

**Discussion by ringohoffman** [microsoft/pyright#7061] [Q&A]
Is there a fundamental difference between the default in `C` and `B` or `A`? I expect the`TypeVar` to bind like it is in `C` in all cases.

```python
class C[X]:
    def __init__(self, x: X = float) -> None:
        self.x = x

c_float = C()  # C[type[float]]
c_str = C(str)  # C[type[str]]

from dataclasses import dataclass

@dataclass
class B[X]:
    x: X = int  # Type "type[int]" cannot be assigned to type "X@B"

b = B()  # B[Unknown]

class A[X]:
    x: X = int  # Type "type[int]" cannot be assigned to type "X@A"

a = A()  # A[Unknown]
```

**Comment by erictraut** [comment]
Yes, these are very different cases.

A generic class is specialized either explicitly (e.g. `C[str]("")`) or implicitly by calling the class constructor with arguments that supply types for the class-scoped type parameters (e.g. `C("")`). If the constructor signature includes default argument values, these can be used instead of arguments that are passed explicitly to the constructor call. That means `C()` is equivalent to `C(float)` in your example above. Both construct a specialized class of type `C[type[float]]`.

In the case of `A`, the constructed type of `A()` is `A[Unknown]` because the constructor signature provides no way to establish the specialized type. The variable `A.x` is a class variable that can be shadowed by an instance variable of the same type, but class variables are shared across all instances of the class, which means that they cannot be generic.

`B` is an interesting case because of the way dataclasses work. Class-scoped variables that are initialized within a dataclass class body are transformed into instance variables by the dataclass decorator, and an `__init__` method is synthesized. The signature of the synthesized `__init__` method looks the same as the one you have manually created for `C`. So this should arguably work without a type error. "Fixing" this would require adding some pretty gross special-case logic in various places, so it's not something I would do unless there's signal that this is an important use case for pyright users in general.

I'll note that mypy also generates errors for all of the above examples.

**Reply by ringohoffman** [reply]
I am asking because I am trying to resolve some type errors in [`nerfstudio`](https://github.com/nerfstudio-project/nerfstudio). Based on how they have structured their config classes, they are running aground of a lot of `reportIncompatibleVariableOverride` and `reportIncompatibleMethodOverride` errors.

This is the general idea of a pattern that is very common in this code base and a general snapshot of the code as it is now.

```python
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

@dataclass
class InstantiateConfig:
    _target: type

    def setup(self, **kwargs: Any) -> Any:
        return self._target(self, **kwargs)

#########################################################

class PixelSampler:
    config: PixelSamplerConfig

@dataclass
class PixelSamplerConfig(InstantiateConfig):
    _target: type[PixelSampler] = PixelSampler

#########################################################

class PatchPixelSampler(PixelSampler):
    config: PatchPixelSamplerConfig  # "config" overrides symbol of same name in class "PixelSampler"

@dataclass
class PatchPixelSamplerConfig(PixelSamplerConfig):
    _target: type[PatchPixelSampler] = PatchPixelSampler  # "_target" overrides symbol of same name in class "PixelSamplerConfig"

#########################################################

class PairPixelSampler(PixelSampler):
    config: PairPixelSamplerConfig  # "config" overrides symbol of same name in class "PixelSampler"

@dataclass
class PairPixelSamplerConfig(PixelSamplerConfig):
    _target: type[PairPixelSampler] = PairPixelSampler  # "_target" overrides symbol of same name in class "PixelSamplerConfig"

```

These errors seem to contradict what you said here: https://github.com/microsoft/pyright/discussions/5320#discussioncomment-6200606

> A subclass can override an attribute type of a parent as long as it is a subtype.

This seems to be the same issue as this discussion https://github.com/microsoft/pyright/discussions/6385

I agree with the sentiment in this post https://github.com/microsoft/pyright/discussions/6872 that I don't want to have to disable `reportIncompatibleVariableOverride` and `reportIncompatibleMethodOverride` if  I have other options, but I am pretty limited with what I can do here.

While mutating these fields is rare, I don't think `nerfstudio` will want to create all their dataclasses with `frozen=True`. There is also no field specific way either in `dataclasses` or `typing` to mark these specific fields as immutable (`typing.Property`?) https://github.com/python/typing/issues/922, https://github.com/python/typing/issues/594.

My last resort is to make these fields' types variable, but the lack of support for binding `TypeVar` on `dataclass` field defaults seems to be blocking my last best effort to resolve these issues.

```python
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

@dataclass
class InstantiateConfig[T]:
    _target: type[T]

    def setup(self, **kwargs: Any) -> T:  # I really want to make this change regardless since more variables will be automatically typed
        return self._target(self, **kwargs)

#########################################################

class PixelSampler[PixelSamplerConfigT: "PixelSamplerConfig"]:
    config: PixelSamplerConfigT

@dataclass
class PixelSamplerConfig[PixelSamplerT: PixelSampler](InstantiateConfig[PixelSamplerT]):
    _target: type[PixelSamplerT] = PixelSampler  # Expression of type "type[PixelSampler]" cannot be assigned to declared type "type[PixelSamplerT@PixelSamplerConfig]"

#########################################################

class PatchPixelSampler[PatchPixelSamplerConfigT: "PatchPixelSamplerConfig"](PixelSampler[PatchPixelSamplerConfigT]):
    config: PatchPixelSamplerConfigT

@dataclass
class PatchPixelSamplerConfig[PatchPixelSamplerT: PatchPixelSampler](PixelSamplerConfig[PatchPixelSamplerT]):
    _target: type[PatchPixelSamplerT] = PatchPixelSampler  # Expression of type "type[PatchPixelSampler]" cannot be assigned to declared type "type[PatchPixelSamplerT@PatchPixelSamplerConfig]"

#########################################################

class PairPixelSampler[PairPixelSamplerConfigT: "PairPixelSamplerConfig"](PixelSampler[PairPixelSamplerConfigT]):
    config: PairPixelSamplerConfigT

@dataclass
class PairPixelSamplerConfig[PairPixelSamplerT: PairPixelSampler](PixelSamplerConfig[PairPixelSamplerT]):
    _target: type[PairPixelSamplerT] = PairPixelSampler  # Expression of type "type[PairPixelSampler]" cannot be assigned to declared type "type[PairPixelSamplerT@PairPixelSamplerConfig]"

```

What would you do in this scenario?

**Reply by erictraut** [reply]
> These errors seem to contradict what you said here...

Some of the discussion threads you're referencing are outdated. After a lengthy debate, the community has concluded that type checkers should [enforce invariance for mutable attributes](https://github.com/microsoft/pyright/issues/5933). That means overrides of mutable attributes must use an equivalent type, not a subtype.

> What would you do in this scenario?

The design pattern used in this code is quite problematic. Dependencies between data structure should be hierarchical, not cyclical. In your example, the "sampler" knows about the "config", and the "config" knows about the "sampler". My recommendation is to refactor the code to eliminate this cycle.
